### PR TITLE
minor refactor to call s2s once for the jobs

### PR DESCRIPTION
**Pull request by iguannago**
https://tools.hmcts.net/jira/browse/RWA-645

### CODE REVIEW

File: src/functionalTest/java/uk/gov/hmcts/reform/wataskmonitor/MonitorTaskJobControllerForAdHocCreateTasksTest.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+import uk.gov.hmcts.reform.wataskmonitor.domain.taskmonitor.request.MonitorTaskJobRequest;
+
+import static net.serenitybdd.rest.SerenityRest.given;
+import static org.hamcrest.Matchers.is;
+import static org.springframework.http.MediaType.APPLICATION_JSON_VALUE;
+import static uk.gov.hmcts.reform.wacaseeventhandler.controllers.MonitorTaskJobControllerUtility.expectedResponse;
+import static uk.gov.hmcts.reform.wataskmonitor.domain.taskmonitor.JobName.AD_HOC_CREATE_TASKS;
+
+@SpringBootTest
+@ActiveProfiles({"local", "functional"})
```
Comment (alexbaeza) [inline src/functionalTest/java/uk/gov/hmcts/reform/wataskmonitor/MonitorTaskJobControllerForAdHocCreateTasksTest.java:26]: Just curious, I noticed this in a few classes is there any reason why you use both profiles, rather than just functional?

File: src/main/java/uk/gov/hmcts/reform/wataskmonitor/clients/IdamWebApi.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    Token token(@RequestBody Map<String, ?> form);
+
+    @Configuration
+    class FeignConfiguration {
+        @Bean
+        @Primary
+        public Encoder feignFormEncoder(ObjectFactory<HttpMessageConverters> messageConverters) {
+            return new SpringFormEncoder(new SpringEncoder(messageConverters));
+        }
+    }
```
Comment (alexbaeza) [inline src/main/java/uk/gov/hmcts/reform/wataskmonitor/clients/IdamWebApi.java:53]: This could be better in its own class under /config as we might need to re-use this.

File: src/main/java/uk/gov/hmcts/reform/wataskmonitor/services/jobs/adhoc/createtasks/ElasticSearchCaseRetrieverService.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+import uk.gov.hmcts.reform.wataskmonitor.domain.jobs.adhoc.createtasks.ElasticSearchRetrieverParameter;
+import uk.gov.hmcts.reform.wataskmonitor.services.jobs.ResourceEnum;
+import uk.gov.hmcts.reform.wataskmonitor.services.jobs.RetrieveCaseListService;
+import uk.gov.hmcts.reform.wataskmonitor.utils.ResourceUtility;
+
+@Component
+public class ElasticSearchCaseRetrieverService implements RetrieveCaseListService<ElasticSearchRetrieverParameter> {
+
+    private final CcdClient ccdClient;
+    private final IdamTokenGenerator systemUserIdamToken;
```
Comment (alexbaeza) [inline src/main/java/uk/gov/hmcts/reform/wataskmonitor/services/jobs/adhoc/createtasks/ElasticSearchCaseRetrieverService.java:17]: This needs to be abstracted into it's own service to implement some caching using the annotations

File: src/main/java/uk/gov/hmcts/reform/wataskmonitor/config/idam/IdamTokenGenerator.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+        map.add("password", userIdamTokenGeneratorInfo.getUserPassword());
+        map.add("scope", userIdamTokenGeneratorInfo.getIdamScope());
+        Token tokenResponse = idamWebApi.token(map);
+
+        return "Bearer " + tokenResponse.getAccessToken();
+    }
+
+    public UserInfo getUserInfo(String bearerAccessToken) {
+        return idamWebApi.userInfo(bearerAccessToken);
+    }
```
Comment (alexbaeza) [inline src/main/java/uk/gov/hmcts/reform/wataskmonitor/config/idam/IdamTokenGenerator.java:37]: We need to cache these two methods :)

File: src/main/java/uk/gov/hmcts/reform/wataskmonitor/services/jobs/adhoc/createtasks/CreateTaskJob.java [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+    @Override
+    @SuppressWarnings("PMD.LawOfDemeter")
+    public boolean canRun(JobName jobName) {
+        return AD_HOC_CREATE_TASKS.equals(jobName);
+    }
+
+    @Override
+    public void run(String serviceToken) {
+        log.info("Starting '{}'", AD_HOC_CREATE_TASKS);
+        log.info("{} finished successfully: {}", AD_HOC_CREATE_TASKS, logPrettyPrint(createTasks(serviceToken)));
```
[the remark this answers is not in the document]
Comment (iguannago) [reply src/main/java/uk/gov/hmcts/reform/wataskmonitor/services/jobs/adhoc/createtasks/CreateTaskJob.java:50]: done

File: src/main/java/uk/gov/hmcts/reform/wataskmonitor/services/jobs/adhoc/createtasks/CreateTaskJobOutcomeService.java [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    public CreateTaskJobOutcomeService(CamundaClient camundaClient) {
+        this.camundaClient = camundaClient;
+    }
+
+    @Override
+    public CreateTaskJobOutcome getJobOutcome(String serviceToken, String caseId) {
+        try {
+            return await()
+                .pollInterval(5, SECONDS)
+                .atMost(15, SECONDS)
```
[the remark this answers is not in the document]
Comment (iguannago) [reply src/main/java/uk/gov/hmcts/reform/wataskmonitor/services/jobs/adhoc/createtasks/CreateTaskJobOutcomeService.java:29]: done

File: src/functionalTest/java/uk/gov/hmcts/reform/wataskmonitor/MonitorTaskJobControllerForAdHocCreateTasksTest.java (hunk 1 above)
Comment (iguannago) [reply src/functionalTest/java/uk/gov/hmcts/reform/wataskmonitor/MonitorTaskJobControllerForAdHocCreateTasksTest.java:26]: hey, yes, to see debug logs

File: src/main/java/uk/gov/hmcts/reform/wataskmonitor/domain/jobs/adhoc/createtasks/ElasticSearchRetrieverParameter.java [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+@Getter
+public class ElasticSearchRetrieverParameter extends RetrieveCaseListParam {
+
+    private final String serviceAuthentication;
+
+    public ElasticSearchRetrieverParameter(String serviceAuthentication) {
+        super();
+        this.serviceAuthentication = serviceAuthentication;
+    }
+}
```
[the remark this answers is not in the document]
Comment (iguannago) [reply src/main/java/uk/gov/hmcts/reform/wataskmonitor/domain/jobs/adhoc/createtasks/ElasticSearchRetrieverParameter.java:19]: When you have to implement the interface, It's nicer not to be forced to implement the method with the same parameters. Therefore, this class shows an example to use different params for different implementations. Yes, it might look redundant now, but I use it to set an example of how to implement the RetrieverService. 
If it becomes useless going further, I'm more than happy to clean up, and I just wanted to put some pattern to make it a bit more flexible.

[the remark this answers is not in the document]
Comment (iguannago) [reply src/main/java/uk/gov/hmcts/reform/wataskmonitor/domain/jobs/adhoc/createtasks/ElasticSearchRetrieverParameter.java:19]: done

File: src/main/java/uk/gov/hmcts/reform/wataskmonitor/services/jobs/adhoc/createtasks/CreateTaskJob.java [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+    }
+
+    @SuppressWarnings("PMD.DataflowAnomalyAnalysis")
+    private CreateTaskJobReport createTasks(String serviceToken) {
+        ElasticSearchCaseList searchCaseList =
+            retrieverService.retrieveCaseList(new ElasticSearchRetrieverParameter(serviceToken));
```
[the remark this answers is not in the document]
Comment (iguannago) [reply src/main/java/uk/gov/hmcts/reform/wataskmonitor/services/jobs/adhoc/createtasks/CreateTaskJob.java]: I think I answered above its benefit to make the interface less rigid and so more flexible. Yes, there are trades-off as always, like the one you mention, more object instance creation. Honestly, I don't think this one has any impact on the performance. Therefore, I'd favour flexibility and leave it as it is for now if we are ok.

[the remark this answers is not in the document]
Comment (iguannago) [reply src/main/java/uk/gov/hmcts/reform/wataskmonitor/services/jobs/adhoc/createtasks/CreateTaskJob.java]: done

File: src/main/java/uk/gov/hmcts/reform/wataskmonitor/services/jobs/adhoc/createtasks/ElasticSearchCaseRetrieverService.java [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+    }
+
+    @Override
+    public ElasticSearchCaseList retrieveCaseList(ElasticSearchRetrieverParameter param) {
+        if (StringUtils.isBlank(param.getServiceAuthentication())) {
+            throw new IllegalArgumentException("service token is missing");
+        }
+        return ccdClient.searchCases(
+            systemUserIdamToken.generate(),
+            param.getServiceAuthentication(),
```
[the remark this answers is not in the document]
Comment (iguannago) [reply src/main/java/uk/gov/hmcts/reform/wataskmonitor/services/jobs/adhoc/createtasks/ElasticSearchCaseRetrieverService.java:31]: Yes, that's another option, Alex. For this case, because the service authentication is generated at the beginning and pass down and reuse among different services, I chose this other option.

[the remark this answers is not in the document]
Comment (iguannago) [reply src/main/java/uk/gov/hmcts/reform/wataskmonitor/services/jobs/adhoc/createtasks/ElasticSearchCaseRetrieverService.java:31]: done

File: src/functionalTest/java/uk/gov/hmcts/reform/wataskmonitor/MonitorTaskJobControllerForAdHocCreateTasksTest.java (hunk 1 above)
Comment (iguannago) [reply src/functionalTest/java/uk/gov/hmcts/reform/wataskmonitor/MonitorTaskJobControllerForAdHocCreateTasksTest.java:26]: done

**Review by alexbaeza — DISMISSED** [review]

File: src/main/java/uk/gov/hmcts/reform/wataskmonitor/services/jobs/adhoc/createtasks/JsonResourceCaseRetrieverService.java [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+import org.springframework.stereotype.Component;
+import uk.gov.hmcts.reform.wataskmonitor.domain.jobs.RetrieveCaseListParam;
+import uk.gov.hmcts.reform.wataskmonitor.domain.jobs.adhoc.createtasks.JsonResourceCaseList;
+import uk.gov.hmcts.reform.wataskmonitor.services.jobs.ResourceEnum;
+import uk.gov.hmcts.reform.wataskmonitor.services.jobs.RetrieveCaseListService;
+import uk.gov.hmcts.reform.wataskmonitor.utils.ObjectMapperUtility;
+import uk.gov.hmcts.reform.wataskmonitor.utils.ResourceUtility;
+
+@Component
+public class JsonResourceCaseRetrieverService implements RetrieveCaseListService<RetrieveCaseListParam> {
```
Comment (BenNzewi) [inline src/main/java/uk/gov/hmcts/reform/wataskmonitor/services/jobs/adhoc/createtasks/JsonResourceCaseRetrieverService.java:12]: JsonResourceCaseRetrieverService - can we add a unit test for this

Comment (iguannago) [reply src/main/java/uk/gov/hmcts/reform/wataskmonitor/services/jobs/adhoc/createtasks/JsonResourceCaseRetrieverService.java:12]: this is covered here ObjectMapperUtilityTest

File: src/test/java/uk/gov/hmcts/reform/wataskmonitor/services/jobs/adhoc/createtasks/CreateTaskJobOutcomeServiceTest.java [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+        );
+
+        Arguments camundaTaskListIsNullScenario = Arguments.of(
+            null,
+            CreateTaskJobOutcome.builder()
+                .caseId(SOME_CASE_ID)
+                .created(false)
+                .build()
+        );
+
```
Comment (BenNzewi) [inline src/test/java/uk/gov/hmcts/reform/wataskmonitor/services/jobs/adhoc/createtasks/CreateTaskJobOutcomeServiceTest.java:127]: As per sonar -> One more edge case test to cover camundaList !empty() Or !=Null And Task.getName().equals("Review Appeal Skeleton Argument") ?

Comment (iguannago) [reply src/test/java/uk/gov/hmcts/reform/wataskmonitor/services/jobs/adhoc/createtasks/CreateTaskJobOutcomeServiceTest.java:138]: done

File: src/main/java/uk/gov/hmcts/reform/wataskmonitor/services/jobs/adhoc/createtasks/JsonResourceCaseRetrieverService.java (hunk 10 above)
Comment (iguannago) [reply src/main/java/uk/gov/hmcts/reform/wataskmonitor/services/jobs/adhoc/createtasks/JsonResourceCaseRetrieverService.java:12]: done

**Review by BenNzewi — APPROVED** [review]

**Review by BenNzewi — APPROVED** [review]
